From 3fe6ae27749b78fad949545a8d6d1a4882ec460c Mon Sep 17 00:00:00 2001
From: Liu Ying <victor.liu@nxp.com>
Date: Fri, 13 Oct 2017 13:22:16 +0800
Subject: [PATCH 2646/5242] MLK-16581-7 drm/imx: ldb: Add system power
 management support

commit  34a3f79a3d83d306103b5fa7f1f65fbaf30f9e4d from
https://source.codeaurora.org/external/imx/linux-imx.git

This patch adds system power management support for imx-ldb drm driver
by proper PHY power/exit/init handling where necessary and pixel link
re-initialization in the resume operation.  The driver depends on the
imx-drm core driver to handle ldb bridge power management operations.

Signed-off-by: Liu Ying <victor.liu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/imx/imx-ldb.c |   38 ++++++++++++++++++++++++++++++++++++++
 1 file changed, 38 insertions(+)

diff --git a/drivers/gpu/drm/imx/imx-ldb.c b/drivers/gpu/drm/imx/imx-ldb.c
index 389d87f..3cc45e5 100644
--- a/drivers/gpu/drm/imx/imx-ldb.c
+++ b/drivers/gpu/drm/imx/imx-ldb.c
@@ -1291,12 +1291,50 @@ static int imx_ldb_remove(struct platform_device *pdev)
 	return 0;
 }
 
+#ifdef CONFIG_PM_SLEEP
+static int imx_ldb_suspend(struct device *dev)
+{
+	struct imx_ldb *imx_ldb = dev_get_drvdata(dev);
+	struct imx_ldb_channel *channel;
+	int i;
+
+	for (i = 0; i < 2; i++) {
+		channel = &imx_ldb->channel[i];
+
+		if (channel->phy_is_on)
+			phy_power_off(channel->phy);
+
+		phy_exit(channel->phy);
+	}
+
+	return 0;
+}
+
+static int imx_ldb_resume(struct device *dev)
+{
+	struct imx_ldb *imx_ldb = dev_get_drvdata(dev);
+	int i;
+
+	if (imx_ldb->visible_phy)
+		for (i = 0; i < 2; i++)
+			phy_init(imx_ldb->channel[i].phy);
+
+	if (imx_ldb->pixel_link_init_quirks)
+		ldb_pixel_link_init(imx_ldb->id);
+
+	return 0;
+}
+#endif
+
+static SIMPLE_DEV_PM_OPS(imx_ldb_pm_ops, imx_ldb_suspend, imx_ldb_resume);
+
 static struct platform_driver imx_ldb_driver = {
 	.probe		= imx_ldb_probe,
 	.remove		= imx_ldb_remove,
 	.driver		= {
 		.of_match_table = imx_ldb_dt_ids,
 		.name	= DRIVER_NAME,
+		.pm	= &imx_ldb_pm_ops,
 	},
 };
 
-- 
1.7.9.5

